### PR TITLE
Secrets file

### PR DESCRIPTION
**Pull request by rishflab**

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -53,3 +55,10 @@ impl Behaviour {
         }
     }
 }
+
+pub fn load_secret_key_from_file(path: impl AsRef<Path>) -> Result<SecretKey> {
+    let contents = fs::read_to_string(path)?;
```
Comment (thomaseizinger) [inline src/lib.rs:60]: `String`s in Rust enforce UTF8. I doubt the bytes of secret keys form valid UTF8 all the time :)

Best to read the content as bytes right away.

Also, best to use tokio's async APIs here. Shouldn't matter much but it is good to develop a habit of not mixing blocking std APIs with async code.

**Review by thomaseizinger — CHANGES_REQUESTED** [review]

**Comment by thomaseizinger** [comment]
Having a flag that generates random bytes and writes them to the given path would be nice. That why getting this thing running would be a bit more convenient.

**Comment by rishflab** [comment]
> Having a flag that generates random bytes and writes them to the given path would be nice. That why getting this thing running would be a bit more convenient.

good idea. what do you think about making the secret file arg optional and if it is not specified, a secret file is generated

**Comment by thomaseizinger** [comment]
> > Having a flag that generates random bytes and writes them to the given path would be nice. That why getting this thing running would be a bit more convenient.
> 
> good idea. what do you think about making the secret file arg optional and if it is not specified, a secret file is generated
> 

Which path will it be generated at if no path is given?

Try to design an interface that:

- is hard to misuse
- yet convenient
- gives the user choice
- is minimal

File: src/lib.rs [hunk 2]
```diff
@@ -53,3 +49,27 @@ impl Behaviour {
...
+
+pub async fn load_secret_key_from_file(path: impl AsRef<Path>) -> Result<SecretKey> {
+    let bytes = fs::read(path).await?;
+    let secret_key = SecretKey::from_bytes(bytes)?;
+    Ok(secret_key)
+}
+
+#[derive(Debug, thiserror::Error)]
+#[error("Could not generate secret file at specified location. File {path} already exists")]
+pub struct FileAlreadyExistsAtSpecifiedLocation {
```
Comment (thomaseizinger) [inline src/lib.rs:61]: Unless you want to downcast to this error somewhere, I would probably not bother making a dedicated type.

You can achieve the same thing with `.context`!

File: src/lib.rs [hunk 3]
```diff
@@ -53,3 +49,27 @@ impl Behaviour {
...
+
+pub async fn generate_secret_key_file(path: PathBuf) -> Result<SecretKey> {
+    if fs::metadata(&path).await?.is_file() {
+        return Err(anyhow::Error::from(FileAlreadyExistsAtSpecifiedLocation {
+            path: path.clone(),
+        }));
+    }
+    let keypair = Keypair::generate();
+    let secret_key = SecretKey::from(keypair);
+    let _ = fs::write(path, secret_key.clone()).await?;
```
Comment (thomaseizinger) [inline src/lib.rs:73]: If you write first, you can avoid a clone because writing only requires a reference (I think).

**Review by thomaseizinger — APPROVED** [review]
Looks good!

Needs a rebase to get rid of the  front commits.

Could sprinkle some `.context` in, depending on how good the error messages currently are when trying to read a file that doesn't exist :)

File: src/lib.rs (hunk 2 above)
Comment (rishflab) [reply src/lib.rs]: removed this. Not needed with the new tokio::fs API's

File: Cargo.toml [hunk 4]
```diff
@@ -9,4 +9,4 @@ anyhow = "1"
 futures = { version = "0.3", default-features = false }
 libp2p = { git = "https://github.com/comit-network/rust-libp2p.git", rev = "4cacaf215b334d447faca8f4e361fe19c30d2162", default-features = false, features = [ "rendezvous", "tcp-tokio", "yamux", "mplex", "dns-tokio", "noise", "ping" ] }
 structopt = { version = "0.3", default-features = false }
-tokio = { version = "1", features = [ "rt-multi-thread", "time", "macros", "sync", "process", "fs", "net" ] }
+tokio = { version = "1.8", features = [ "rt-multi-thread", "time", "macros", "sync", "process", "fs", "net", "io-util" ] }
```
Comment (thomaseizinger) [inline Cargo.toml:12]: `1` implies `1.8`. Why was this change necessary? Familiarize yourself with `cargo update` if you want the change the versions in the lockfile.

File: src/bin/rendezvous_server.rs [hunk 5]
```diff
@@ -1,28 +1,38 @@
...
+use std::path::PathBuf;
 use structopt::StructOpt;
 
 #[derive(Debug, StructOpt)]
 struct Cli {
-    #[structopt(long = "secret-key", help = "32 byte string", parse(try_from_str = parse_secret_key))]
-    pub secret_key: SecretKey,
+    #[structopt(
+        long = "secret-file",
+        help = "Path to the file that contains the secret used to derive the rendezvous server's identity"
```
Comment (thomaseizinger) [inline src/bin/rendezvous_server.rs:17]: Nit: I don't see any _derivation_ anywhere in the code.

File: src/lib.rs [hunk 6]
```diff
@@ -53,3 +52,37 @@ impl Behaviour {
         }
     }
 }
+
+pub async fn load_secret_key_from_file(path: impl AsRef<Path> + Debug) -> Result<SecretKey> {
+    let bytes = fs::read(&path)
+        .await
+        .context(format!("No secret file at {:?}", path))?;
```
Comment (thomaseizinger) [inline src/lib.rs:59]: Have a look at anyhow's docs regarding `.context` and `with_context`: https://docs.rs/anyhow/1.0.41/anyhow/trait.Context.html#required-methods

Then, explain why you are using `context` here and not `with_context`.

File: src/lib.rs [hunk 7]
```diff
@@ -53,3 +52,37 @@ impl Behaviour {
...
+pub async fn generate_secret_key_file(path: PathBuf) -> Result<SecretKey> {
+    if let Some(parent) = path.parent() {
+        DirBuilder::new()
+            .recursive(true)
+            .create(parent)
+            .await
+            .context(format!(
+                "Could not create directory for secret file: {:?}",
+                parent
+            ))?;
```
Comment (thomaseizinger) [inline src/lib.rs:73]: [`tokio::fs::create_dir_all`](https://docs.rs/tokio/1.8.0/tokio/fs/fn.create_dir_all.html) is quite a bit shorter no? :)

**Review by thomaseizinger — CHANGES_REQUESTED** [review]
Front commit is already (partly) in master no? I think that needs to be dropped.

Remaining comments are not blocking.

**Comment by rishflab** [comment]
> Front commit is already (partly) in master no? I think that needs to be dropped.

nope. the commit in master adds ARM to the ci.yml where as this commit adds it to the build-release-binaries.yml

File: src/lib.rs [hunk 8]
```diff
@@ -56,7 +56,7 @@ impl Behaviour {
 pub async fn load_secret_key_from_file(path: impl AsRef<Path> + Debug) -> Result<SecretKey> {
     let bytes = fs::read(&path)
         .await
-        .context(format!("No secret file at {:?}", path))?;
+        .with_context(|| format!("No secret file at {:?}", path))?;
```
Comment (thomaseizinger) [inline src/lib.rs:59]: Paths have a `Display` adapter: https://doc.rust-lang.org/std/path/struct.Path.html#method.display

**Review by thomaseizinger — APPROVED** [review]
Good to be merged :)
